openmdao/core/problem.py

```diff
@@ -1890,40 +1936,66 @@ def _assemble_derivative_data(derivative_data, rel_error_tol, abs_error_tol, out
                         error_descs = ('(Jfor  - Jfd) ', '(Jrev  - Jfd) ', '(Jfor  - Jrev)')
                     for error, desc in zip(abs_err, error_descs):
                         error_str = _format_error(error, abs_error_tol)
+                        if error_str.endswith('*'):
+                            num_bad_jacs += 1
                         if out_stream:
-                            out_stream.write('    Absolute Error {}: {}'.format(desc, error_str) +
+                            out_buffer.write('    Absolute Error {}: {}'.format(desc, error_str) +
                                              '\n')
                     if out_stream:
-                        out_stream.write('\n')
+                        out_buffer.write('\n')
 
                     # Relative Errors
                     for error, desc in zip(rel_err, error_descs):
                         error_str = _format_error(error, rel_error_tol)
+                        if error_str.endswith('*'):
+                            num_bad_jacs += 1
                         if out_stream:
-                            out_stream.write('    Relative Error {}: {}'.format(desc, error_str) +
+                            out_buffer.write('    Relative Error {}: {}'.format(desc, error_str) +
                                              '\n')
+
                     if out_stream:
-                        out_stream.write('\n')
+                        out_buffer.write('\n')
 
                     # Raw Derivatives
                     if out_stream:
-                        out_stream.write('    Raw Forward Derivative (Jfor)\n')
-                        out_stream.write(str(forward) + '\n')
-                        out_stream.write('\n')
+                        out_buffer.write('    Raw Forward Derivative (Jfor)\n')
+                        out_buffer.write(str(forward) + '\n')
+                        out_buffer.write('\n')
 
                     if not totals and system.matrix_free:
                         if out_stream:
-                            out_stream.write('    Raw Reverse Derivative (Jfor)\n')
-                            out_stream.write(str(reverse) + '\n')
-                            out_stream.write('\n')
+                            out_buffer.write('    Raw Reverse Derivative (Jfor)\n')
+                            out_buffer.write(str(reverse) + '\n')
+                            out_buffer.write('\n')
 
                     if out_stream:
-                        out_stream.write('    Raw FD Derivative (Jfd)\n')
-                        out_stream.write(str(fd) + '\n')
-                        out_stream.write('\n')
+                        out_buffer.write('    Raw FD Derivative (Jfd)\n')
+                        out_buffer.write(str(fd) + '\n')
+                        out_buffer.write('\n')
 
                     if out_stream:
-                        out_stream.write(' -' * 30 + '\n')
+                        out_buffer.write(' -' * 30 + '\n')
```

Comment:
could some of these `if out_stream:` blocks be consolidated?  for example, it looks like lines 1956 through 1977 could be handled with a single such conditional rather than five...

Reply:
They sure could! There's a story to re-factor these methods. Included in that is getting rid of things like the suppress_output option and things like you are talking about. Really need to reduce the if's and reduce the depths of the blocks.